src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```
@@ -1,25 +1,108 @@
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private final List<Product> products;
+    private long currentId;
+    private final ReentrantReadWriteLock locker = new ReentrantReadWriteLock();
+
+    public ArrayListProductDao() {
+        products = new ArrayList<>();
+        saveSampleProducts();
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public Optional<Product> getProduct(Long id) {
+        locker.readLock().lock();
+        try {
+            return products.stream()
+                    .filter(product -> id.equals(product.getId()))
+                    .findAny();
+        } finally {
+            locker.readLock().unlock();
+        }
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        locker.readLock().lock();
+        try {
+            return products.stream()
+                    .filter(this::productHasNonNullPrice)
+                    .filter(this::productIsInStock)
+                    .collect(Collectors.toList());
+        } finally {
+            locker.readLock().unlock();
+        }
+    }
+
+    private boolean productHasNonNullPrice(Product product) {
+        return product.getPrice() != null;
+    }
+
+    private boolean productIsInStock(Product product) {
+        return product.getStock() > 0;
     }
 
     @Override
     public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+        locker.writeLock().lock();
+        try {
+            if (product.getId() == null) {
+                product.setId(++currentId);
+                products.add(product);
+            } else {
```

Don't split else and if like here. You can define it in one line - else if () {

---

The code in this else clause can be simplified. With the next alrogithm:

1) if product with given id exists then update it
2) if getProduct(id) for given id returns Optional.EMPTY then you can throw an exception